ts/src/flexible-event/privacy.ts

```diff
@@ -123,46 +129,55 @@ export function flipProbabilityDp(numStates: number, epsilon: number): number {
   return numStates / (numStates + Math.exp(epsilon) - 1)
 }
 
-/**
- * Computes the capacity of the q-ary symmetric channel.
- *
- * @param log2q - the logarithm to base 2 of the alphabet size.
- * @param flipProbability - the channel keeps the input the same with probability
- *   1 - flipProbability, and flips the input to one of the other q - 1
- *   symbols (uniformly) with the remaining probability of flipProbability.
- *
- * @returns - The capacity of the q-ary symmetric channel for given flipProbability.
- *   In general, the capacity is defined as the maximum, over all input
- *   distributions, of the mutual information between the input and output of
- *   the channel. In the special case of the q-ary symmetric channel, a
- *   closed-form expression is known, which we use here.
- */
-function capacityQarySymmetricChannel(
-  log2q: number,
-  flipProbability: number
-): number {
-  return (
-    log2q -
-    binaryEntropy(flipProbability) -
-    flipProbability * Math.log2(Math.pow(2, log2q) - 1)
-  )
+// Returns the pick rate for a source with attribution scope.
+export function fakeProbabilityDp(
+  numStates: number,
+  numUsedScopes : number,
+  numEventStates: number, 
+  epsilon: number): number {
+  const pickRateForSource = flipProbabilityDp(numStates, epsilon) * (numStates - 1) / numStates
+  const pickRateForEvent = flipProbabilityDp(numEventStates, epsilon) * (numEventStates- 1) / numEventStates;
+  return 1 - (1 - pickRateForSource) * Math.pow(1 - pickRateForEvent, numUsedScopes)
 }
 
-export function maxInformationGain(numStates: number, epsilon: number): number {
-  const flipProb = flipProbabilityDp(numStates, epsilon)
-  if (numStates === 1 || flipProb === 1) {
+export function informationGain(
+  numStates: number,
+  numUsedScopes: number,
+  numEventStates: number,
+  epsilon: number): number {
+  // Use BigInt for the total number of states to avoid potential overflow.
+  const totalNumStates = BigInt(numStates) + BigInt(numEventStates) * BigInt(numUsedScopes)
+  if (totalNumStates <= 1n) {
     return 0
   }
-  return capacityQarySymmetricChannel(
-    Math.log2(numStates),
-    (flipProb * (numStates - 1)) / numStates
-  )
+  const log2Q = Math.log2(Number(totalNumStates))
+  const fakeProbability = fakeProbabilityDp(numStates, numUsedScopes, numEventStates, epsilon)
+  return (
+    log2Q - 
+    binaryEntropy(fakeProbability) -
+    fakeProbability * Math.log2(Number(totalNumStates) - 1))
+}
+
+export function maxInformationGain(
+  numStates: number,
+  epsilon: number,
+  attributionScopeLimit?: number | null,
+  maxEventStates?: number | null): number {
+  let maxInfoGain = 0
+  for (let k = 0; k < (attributionScopeLimit || 1); ++k) {
+    for (let q2 = 1; q2 <= (maxEventStates || constants.defaultMaxEventStates); ++q2) {
```

Comment:
0 is falsy, so `||` may not do what we want here. Please use [`??`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing) instead.

Reply:
Fixed in the callers.